Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -324,6 +324,12 @@ public struct ByteBuffer {
             self.bytes = ptr
             self.capacity = newCapacity
         }
+        
+        @inlinable
+        func clampStorage(to capacity: Int) {
+            self.bytes.copyMemory(from: self.bytes, byteCount: capacity)
+            self.capacity = _toCapacity(capacity)
```

Comment:
This doesn't effectively do what we want: it shrinks what we _report_ the capacity to be, but not what it _is_. Clamping the storage will almost always require a new allocation.

As we currently use a power-of-two capacity policy, we should probably only actually do the reallocation when the `desiredCapacity` is smaller than the current capacity _after_ applying the next power of two math. At that point, we should likely just call `reallocSlice` on the storage.

Reply:
I have pushed a commit. I think it addresses what you have said here. Please take a look. I removed the _copyStorageAndRebase optimisation since it seems to be doing exactly the same thing, which was reallocSlice.